File: FileType/FileType.cpp
// FileType.cpp

#include "StdAfx.h"
#include ".\FileType.h"


namespace UniversalFileExplorer
{
	
	// Constructor
	FileType::FileType(void)
	{		
		m_debug = gcnew UFEDebug();

		m_DebugText = "";
		
		// Initialize FileCheckInfo
		m_fileCheckInfo = gcnew FileCheckInfo();

		// Create the ArrayToNum instance
		m_atn = gcnew ArrayToNum();

		// Create an instance of the NumToString Class
		m_nts = gcnew NumToString();

		// Set the number format for the NumToString
		m_nts->SetNumFormat(NumberFormat::Default);

		m_Icons = gcnew ArrayList();
		
		m_useIcons = false;
		//LoadIcons();
		
		// By default these tabs are hidden
		m_showTechnical = false;
		m_showGraphic = false;
		m_showFileCheck = false;

		m_rootTreeNode = gcnew TreeNode("ROOT");
	}

	// Deconstructor
	FileType::~FileType(void)
	{		
		m_DebugText = nullptr;

		// Free the file stream
		if(m_FileStream != nullptr)
		{
			m_FileStream->Close();
			delete m_FileStream;
			m_FileStream = nullptr;
		}

		// Delete the trees
		if(m_rootTreeNode != nullptr)
		{
			//m_rootTreeNode->Nodes->Clear();
			m_rootTreeNode = nullptr;
		}
	}
	
	void FileType::DebugOut(String^ NewText)
	{
		if(!m_DebugText->Equals(""))
			m_DebugText = String::Concat(m_DebugText, "\r\n", NewText);
		else
			m_DebugText = NewText;

		m_debug->NewInfo(NewText);
	}

	// ExceptionOut(Exception* e)
	//		Adds an exception to the DebugInfo
	void FileType::ExceptionOut(Exception^ e)
	{
		m_debug->NewException(e);
	}
	
	void FileType::ExceptionOut(Exception^ e, String^ className)
	{
		m_debug->NewException(e, className);
	}

	void FileType::ExceptionOut(Exception^ e, String^ className, String^ funcName)
	{
		m_debug->NewException(e, className, funcName);
	}
/*
	String* FileType::GetUniStr(unsigned char asciiBytes __gc[])
	{	
		int numNullChars = 0;
		Byte asciiBytes2 __gc[];
		bool moreNullChars = true;

		while(moreNullChars)
		{
			if((asciiBytes->Count - 1 - numNullChars) >= 0)
			{
				if(asciiBytes[asciiBytes->Count - 1 - numNullChars] == 0x00)
					numNullChars++;
				else
					moreNullChars = false;
			}
			else
				moreNullChars = false;
		}

		asciiBytes2 = new Byte[asciiBytes->Count - numNullChars];

		for(int i = 0; i < asciiBytes2->Count; i++)
		{
			asciiBytes2[i] = asciiBytes[i];
		}
		
		String* unicodeString = S"";

		// Create two different encodings.
		Encoding* ascii = Encoding::ASCII;
		Encoding* unicode = Encoding::Unicode;
		
		// Perform the conversion from one encoding to the other.
		Byte unicodeBytes[] = Encoding::Convert(ascii, unicode, asciiBytes2);
		asciiBytes2->Finalize();
		asciiBytes2 = NULL;
		ascii->Finalize();
		ascii = NULL;
		Char unicodeChars[] = new Char[unicode->GetCharCount(unicodeBytes, 0, unicodeBytes -> Length)];
		unicode->GetChars(unicodeBytes, 0, unicodeBytes->Length, unicodeChars, 0);
		unicodeBytes->Finalize();
		unicodeBytes = NULL;
		unicode->Finalize();
		unicode = NULL;
		unicodeString = new String(unicodeChars);
		unicodeChars->Finalize();
		unicodeChars = NULL;
		return unicodeString;
	}
	*/
	void FileType::LoadIcons()
	{
		/*
		String* iconNames[] = {S"Null.ico", S"Circle.ico", S"Square.ico",
							 S"Table.ico", S"Properties.ico",
							 S"Text.ico", S"Script.ico",
							 S"Book.ico",
							 S"Objects.ico",
							 S"FolderOpen.ico", S"FolderClosed.ico"};
		*/
		array<String^>^ iconNames = gcnew array<String^>(11);
		iconNames[0] = "Null.ico";
		iconNames[1] = "Circle.ico";
		iconNames[2] = "Square.ico";
		iconNames[3] = "Table.ico";
		iconNames[4] = "Properties.ico";
		iconNames[5] = "Text.ico";
		iconNames[6] = "Script.ico";
		iconNames[7] = "Book.ico";
		iconNames[8] = "Objects.ico";
		iconNames[9] = "FolderOpen.ico";
		iconNames[10] = "FolderClosed.ico";

		
		ResourceManager^ resourceManager = gcnew ResourceManager("FileType.ResourceFiles", Assembly::GetExecutingAssembly());
		try
		{
			for (int i = 0; i < iconNames->Length; i++) {
				AddIcon(static_cast<Icon^>(resourceManager->GetObject(iconNames[i])));
				DebugOut(iconNames[i]);
			}
		}
		catch(Exception^ e)
		{
			DebugOut(e->Message);
			m_useIcons = false;
		}
	}

	bool FileType::ProcessFile()
	{
		return true;
	}
	
	TableData^ FileType::GetData(TreeNode^ tn)
	{
		return nullptr;
	}

	QuickInfoTable^ FileType::GetQuickInfo()
	{
		return nullptr;
	}

	Image^ FileType::GetImage()
	{
		return nullptr;
	}

	String^ FileType::GetFileCreator()
	{
		return "";
	}

};

File: UfexAPI/FileType.cs
using System;
using System.Collections;
using System.Resources;
using System.Drawing;
using System.Windows.Forms;
using System.Reflection;
using UniversalFileExplorer;
using System.IO;

namespace Ufex.API
{
	public abstract class FileType
	{
		public String m_DebugText;
		public String m_AppPath;
		public FileStream m_FileStream;
		private String m_filePath;

		private TreeNode m_rootTreeNode;

		private String m_description;

		private NumberFormat m_numFormat;

		private FileCheckInfo m_fileCheckInfo;
		
		private ArrayToNum m_atn;
		private DataFormatter m_nts;

		private ArrayList m_Icons;
		
		private bool m_useIcons;

		private String m_appPath;

		// Tabs to be displayed
		private bool m_showTechnical;
		private bool m_showGraphic;
		private bool m_showFileCheck;

		// 
		private Logger m_debug;

		public FileType()
		{
			m_debug = new Logger();
			m_DebugText = "";

			// Initialize FileCheckInfo
			m_fileCheckInfo = new FileCheckInfo();

			// Create the ArrayToNum instance
			m_atn = new ArrayToNum();

			// Create an instance of the NumToString Class
			m_nts = new DataFormatter();

			// Set the number format for the NumToString
			m_nts.NumFormat = NumberFormat.Default;

			m_Icons = new ArrayList();

			m_useIcons = false;
			//LoadIcons();

			// By default these tabs are hidden
			m_showTechnical = false;
			m_showGraphic = false;
			m_showFileCheck = false;

			m_rootTreeNode = new TreeNode("ROOT");
		}

		~FileType()
		{
			m_DebugText = null;

			// Free the file stream
			if (m_FileStream != null)
			{
				m_FileStream.Close();
				m_FileStream = null;
			}

			// Delete the trees
			if (m_rootTreeNode != null)
			{
				//m_rootTreeNode.Nodes.Clear();
				m_rootTreeNode = null;
			}
		}

		public String FilePath 
		{
			get { return this.m_filePath; }
			set { this.m_filePath = value; }
		}

		public FileStream FileInStream {
			get { return m_FileStream; }
		}

		public NumberFormat NumFormat 
		{
			set { m_nts.SetNumFormat(value); }
		}

		public TreeNodeCollection TreeNodes 
		{
			get { return m_rootTreeNode.Nodes; }
		}

		public FileCheckInfo FileCheck 
		{
			get { return m_fileCheckInfo; }
		}

		protected DataFormatter NTS 
		{
			get { return m_nts; }
		}

		protected ArrayToNum ATN 
		{
			get { return m_atn; }
		}

		public String ApplicationPath 
		{
			get { return m_appPath; }
		}

		public String Description 
		{
			get { return m_description; }
			protected set { m_description = value; }
		}

		public Boolean UseTreeViewIcons 
		{
			get { return m_useIcons; }
			protected set { m_useIcons = value; }
		}
		
		// Determines whether the Technical View Tab is displayed
		public Boolean ShowTechnical 
		{
			get { return m_showTechnical; }
			protected set { m_showTechnical = value; }
		}
		
		public Boolean ShowGraphic 
		{
			get { return m_showGraphic; }
			protected set { m_showGraphic = value; }
		}

		public Boolean ShowFileCheck 
		{
			get { return m_showFileCheck; }
			protected set { m_showFileCheck = value; }
		}

		[Obsolete("Use Log instead")]
		protected Logger Debug 
		{
			get { return m_debug; }
		}

		protected Logger Log
		{
			get { return m_debug; }
		}

		abstract public bool ProcessFile();

		// Returns the data table that corresponds to the treenode
		virtual public Ufex.API.Tables.TableData GetData(TreeNode tn)
		{
			return null;
		}

		// Returns the Quick Info for the file
		virtual public Ufex.API.Tables.QuickInfoTableData GetQuickInfo()
		{
			return null;
		}

		// Returns the File's Creator
		virtual public String GetFileCreator()
		{
			return "";
		}

		// Returns an image representation of the file
		virtual public Image GetImage()
		{
			return null;
		}

		// Add an icon - returns the icon id number
		public int AddIcon(Icon ico) 
		{ 
			return m_Icons.Add(ico); 
		}

		public int GetNumIcons() 
		{
			return m_Icons.Count; 
		}

		public Icon GetIcon(int i) { 
			return (Icon)m_Icons[i]; 
		}

		// Functions for interfacing with the ArrayToNum Class
		public void SetATNEndian(Endian endian) 
		{ 
			m_atn.DataEndian = endian == Endian.Little ? UniversalFileExplorer.Endian.Little : UniversalFileExplorer.Endian.Big; 
		}

		// ExceptionOut(Exception* e)
		//		Adds an exception to the DebugInfo
		protected void ExceptionOut(Exception e)
		{
			m_debug.NewException(e);
		}

		protected void ExceptionOut(Exception e, String className)
		{
			m_debug.NewException(e, className);
		}

		protected void ExceptionOut(Exception e, String className, String funcName)
		{
			m_debug.NewException(e, className, funcName);
		}

		protected void DebugOut(String NewText)
		{
			if (!m_DebugText.Equals(""))
				m_DebugText = m_DebugText + "\r\n" + NewText;
			else
				m_DebugText = NewText;

			m_debug.Info(NewText);
		}

		private void LoadIcons()
		{
			string[] iconNames = {
				"Null.ico",
				"Circle.ico",
				"Square.ico",
				"Table.ico",
				"Properties.ico",
				"Text.ico",
				"Script.ico",
				"Book.ico",
				"Objects.ico",
				"FolderOpen.ico",
				"FolderClosed.ico"
			};

			ResourceManager resourceManager = new ResourceManager("FileType.ResourceFiles", Assembly.GetExecutingAssembly());
			try
			{
				for (int i = 0; i < iconNames.Length; i++)
					AddIcon((Icon)resourceManager.GetObject(iconNames[i]));
			}
			catch (Exception e)
			{
				this.DebugOut(e.Message);
				m_useIcons = false;
			}
		}

	}
}
